keps/sig-apps/3715-elastic-indexed-job/README.md

```diff
@@ -0,0 +1,936 @@
+  - [Make spec.completions fully mutable](#make-speccompletions-fully-mutable)
+  - [Allow spec.completions=nil for Indexed mode](#allow-speccompletionsnil-for-indexed-mode)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+<!--
+**ACTION REQUIRED:** In order to merge code into a release, there must be an
+issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [x] (R) Design details are appropriately documented
+- [x] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [x] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Currently `spec.completions` is an immutable in Job. This KEP proposes a new
+`CompletionMode` named `ElasticIndexed` that behaves similar to `Indexed` Job,
+but allows to mutate both `spec.completions` and `spec.parallelsim` if done together.
```

Comment:
The risks for mutating `.spec.completions` still hold and based on the discussions we've had so far I don't see strong arguments towards relaxing that. How do you expect a controller to react if it's just marking a job as completed, while a user at the same time modifies completions?

Reply:
> How do you expect a controller to react if it's just marking a job as completed, while a user at the same time modifies completions?

We're already relying on the workloads in an autoscaling job to be coherent and not complete with success until the work is actually done, so I'm not sure the "appears complete" and "needs to scale up" race would happen much in practice, but for correctness, I would expect server validation and controller preconditions to prevent the race:
* controller marking complete would get a resourceVersion conflict error on write if the user had since mutated completions/parallelism
* user mutating completions/parallelism would be rejected if the job is already complete